<a href="https://colab.research.google.com/github/mka-2000/Excelr-assignment/blob/main/Random_forest(Assignment1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# We are given with the fraud data .

# Using Random Forest technique , we need to classify the dataset.

In [2]:
# Importing Libraries

In [3]:
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier

In [4]:
from google.colab import files
uploaded = files.upload()

Saving Fraud_check (1).csv to Fraud_check (1).csv


In [6]:
fc = pd.read_csv('Fraud_check (1).csv') 

In [7]:
fc

Undergrad Marital.Status  Taxable.Income  City.Population  \
0          NO         Single           68833            50047   
1         YES       Divorced           33700           134075   
2          NO        Married           36925           160205   
3         YES         Single           50190           193264   
4          NO        Married           81002            27533   
..        ...            ...             ...              ...   
595       YES       Divorced           76340            39492   
596       YES       Divorced           69967            55369   
597        NO       Divorced           47334           154058   
598       YES        Married           98592           180083   
599        NO       Divorced           96519           158137   

     Work.Experience Urban  
0                 10   YES  
1                 18   YES  
2                 30   YES  
3                 15   YES  
4                 28    NO  
..               ...   ...  
595                7   YES  
596                2   YES  
597                0   YES  
598               17    NO  
599               16    NO  

[600 rows x 6 columns]

Label Encoder for Categorical Datas

In [8]:
from sklearn.preprocessing import LabelEncoder

labelencoder = LabelEncoder()

fc['Undergrad1'] = labelencoder.fit_transform(fc['Undergrad'])
fc['Marital.Status1'] = labelencoder.fit_transform(fc['Marital.Status'])
fc['Urban1'] = labelencoder.fit_transform(fc['Urban'])

In [9]:
fc

Undergrad Marital.Status  Taxable.Income  City.Population  \
0          NO         Single           68833            50047   
1         YES       Divorced           33700           134075   
2          NO        Married           36925           160205   
3         YES         Single           50190           193264   
4          NO        Married           81002            27533   
..        ...            ...             ...              ...   
595       YES       Divorced           76340            39492   
596       YES       Divorced           69967            55369   
597        NO       Divorced           47334           154058   
598       YES        Married           98592           180083   
599        NO       Divorced           96519           158137   

     Work.Experience Urban  Undergrad1  Marital.Status1  Urban1  
0                 10   YES           0                2       1  
1                 18   YES           1                0       1  
2                 30   YES           0                1       1  
3                 15   YES           1                2       1  
4                 28    NO           0                1       0  
..               ...   ...         ...              ...     ...  
595                7   YES           1                0       1  
596                2   YES           1                0       1  
597                0   YES           0                0       1  
598               17    NO           1                1       0  
599               16    NO           0                0       0  

[600 rows x 9 columns]

In [11]:
ti_check = []
for i in fc['Taxable.Income']:
  if i<= 30000:
    print('risky')
  else:
    print('good') 
     

good
good
good
good
good
good
good
good
good
good
risky
good
risky
good
good
good
risky
good
good
risky
good
risky
good
good
good
good
good
good
risky
good
good
good
good
good
risky
good
risky
good
risky
good
good
good
good
good
good
good
good
good
risky
good
good
good
good
risky
good
risky
good
good
risky
good
good
risky
good
good
good
good
good
good
risky
good
good
good
good
good
good
good
good
good
good
risky
good
good
good
good
good
risky
good
risky
good
risky
good
good
good
risky
risky
good
risky
good
risky
good
good
good
good
good
good
good
risky
good
good
good
risky
good
good
good
risky
good
good
good
risky
good
good
risky
risky
risky
good
good
risky
good
good
good
good
good
good
good
good
risky
good
risky
good
good
risky
good
good
risky
good
good
good
good
good
good
good
good
risky
risky
risky
good
good
good
risky
good
good
good
good
risky
good
good
good
good
risky
good
good
risky
risky
good
good
good
good
good
good
risky
good
risky
risky
good
good
good
good
good
risky
good
goo

In [12]:
fc['Taxable.Income']=pd.cut(fc['Taxable.Income'],bins=[0,30000,100000],labels=['risky','good'])

In [13]:
fc

Undergrad Marital.Status Taxable.Income  City.Population  Work.Experience  \
0          NO         Single           good            50047               10   
1         YES       Divorced           good           134075               18   
2          NO        Married           good           160205               30   
3         YES         Single           good           193264               15   
4          NO        Married           good            27533               28   
..        ...            ...            ...              ...              ...   
595       YES       Divorced           good            39492                7   
596       YES       Divorced           good            55369                2   
597        NO       Divorced           good           154058                0   
598       YES        Married           good           180083               17   
599        NO       Divorced           good           158137               16   

    Urban  Undergrad1  Marital.Status1  Urban1  
0     YES           0                2       1  
1     YES           1                0       1  
2     YES           0                1       1  
3     YES           1                2       1  
4      NO           0                1       0  
..    ...         ...              ...     ...  
595   YES           1                0       1  
596   YES           1                0       1  
597   YES           0                0       1  
598    NO           1                1       0  
599    NO           0                0       0  

[600 rows x 9 columns]

In [15]:
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()
fc['TI'] = labelencoder.fit_transform(fc['Taxable.Income'])

In [16]:
fc

Undergrad Marital.Status Taxable.Income  City.Population  Work.Experience  \
0          NO         Single           good            50047               10   
1         YES       Divorced           good           134075               18   
2          NO        Married           good           160205               30   
3         YES         Single           good           193264               15   
4          NO        Married           good            27533               28   
..        ...            ...            ...              ...              ...   
595       YES       Divorced           good            39492                7   
596       YES       Divorced           good            55369                2   
597        NO       Divorced           good           154058                0   
598       YES        Married           good           180083               17   
599        NO       Divorced           good           158137               16   

    Urban  Undergrad1  Marital.Status1  Urban1  TI  
0     YES           0                2       1   0  
1     YES           1                0       1   0  
2     YES           0                1       1   0  
3     YES           1                2       1   0  
4      NO           0                1       0   0  
..    ...         ...              ...     ...  ..  
595   YES           1                0       1   0  
596   YES           1                0       1   0  
597   YES           0                0       1   0  
598    NO           1                1       0   0  
599    NO           0                0       0   0  

[600 rows x 10 columns]

In [17]:
# Seperating independent and dependent variable

#  x  - independent variable
#  y  - dependent variable

In [19]:
x = fc.iloc[:,[3,4,6,7,8]]
y = fc.iloc[:,-1]

In [20]:
# Bagging

In [23]:
kfold = KFold(n_splits=10,random_state=7,shuffle=True)
cart = DecisionTreeClassifier()
model = BaggingClassifier(base_estimator=cart,n_estimators=200,random_state=7)
results = cross_val_score(model,x,y,cv=kfold)
results.mean()

0.7416666666666666

In [24]:
# Random Forest Classification

In [25]:
from sklearn.ensemble import RandomForestClassifier
kfold = KFold(n_splits=10,random_state=7,shuffle=True)
model = RandomForestClassifier(n_estimators=1000,max_features=5,criterion='gini')

results = cross_val_score(model,x,y,cv=kfold)

results.mean()

0.7449999999999999

In [26]:
# Adaboost Classification

In [28]:
from sklearn.ensemble import AdaBoostClassifier

kfold = KFold(n_splits=10,random_state=7,shuffle=True)
model = AdaBoostClassifier(n_estimators=10,random_state=7)
results = cross_val_score(model,x,y,cv=kfold)
results.mean()

0.7933333333333332

In [29]:
# Stacking Ensemble estimators for Classification

In [30]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier


In [31]:
kfold = KFold(n_splits=10,random_state=7,shuffle = True)

In [32]:
# Create the sub models

In [33]:
estimators = []

model1 = LogisticRegression(max_iter=500)
estimators.append(('logistic',model1))
model2 = DecisionTreeClassifier()
estimators.append(('cart',model2))
model3 = SVC()
estimators.append(('svm',model3))

In [36]:
estimators

[('logistic', LogisticRegression(max_iter=500)),
 ('cart', DecisionTreeClassifier()),
 ('svm', SVC())]

In [34]:
# Creating Ensemble Model

In [40]:
ensemble = VotingClassifier(estimators)
results = cross_val_score(ensemble,x,y,cv=kfold)
print(results.mean())
estimators

0.7933333333333332


[('logistic', LogisticRegression(max_iter=500)),
 ('cart', DecisionTreeClassifier()),
 ('svm', SVC())]

Conclusion:

The mean of the Ensemble model comes out to be 0.7933.

